src/Utilities/model.jl

```diff
@@ -168,48 +168,70 @@ function check_can_assign_name(model::MOI.ModelLike, IndexType::Type{<:MOI.Index
 end
 
 """
-    setname(index_to_names::Dict{<:MOI.Index, String}, names_to_index::Dict{String, <:MOI.Index}, idx::MOI.Index, name::String, idxtype::Symbol)
+    setname(index_to_name::Dict{<:MOI.Index, String}, name_to_index::Dict{String, <:MOI.Index}, idx::MOI.Index, name::String, idxtype::Symbol)
 
-Sets the name of the index `idx` to the name `name` in the maps `index_to_names` and `names_to_index`.
+Sets the name of the index `idx` to the name `name` in the maps `index_to_name` and `name_to_index`.
 If `name` is empty, and `idx` already has a name, it is removed.
 """
-function setname(index_to_names::Dict{<:MOI.Index, String}, names_to_index::Dict{String, <:MOI.Index}, idx::MOI.Index, name::String)
-    if haskey(index_to_names, idx)
-        delete!(names_to_index, index_to_names[idx])
+function setname(index_to_name::Dict{<:MOI.Index, String}, name_to_index::Dict{String, <:MOI.Index}, idx::MOI.Index, name::String)
+    if haskey(index_to_name, idx)
+        delete!(name_to_index, index_to_name[idx])
     end
-    index_to_names[idx] = name
+    index_to_name[idx] = name
     if !isempty(name)
-        names_to_index[name] = idx
+        name_to_index[name] = idx
     end
     return
 end
 
 MOI.supports(::AbstractModel, ::MOI.VariableName, vi::Type{VI}) = true
 function MOI.set(model::AbstractModel, ::MOI.VariableName, vi::VI, name::String)
-    if check_can_assign_name(model, VI, vi, name)
-        setname(model.varnames, model.namesvar, vi, name)
-    end
+    model.var_to_name[vi] = name
+    model.name_to_var = nothing # Invalidate the name map.
 end
-MOI.get(model::AbstractModel, ::MOI.VariableName, vi::VI) = get(model.varnames, vi, EMPTYSTRING)
+MOI.get(model::AbstractModel, ::MOI.VariableName, vi::VI) = get(model.var_to_name, vi, EMPTYSTRING)
 
 function MOI.get(model::AbstractModel, ::Type{VI}, name::String)
-    return get(model.namesvar, name, nothing)
+    if model.name_to_var === nothing
+        # Rebuild the map.
+        model.name_to_var = Dict{String, VI}()
+        for (var, var_name) in model.var_to_name
+            isempty(var_name) && continue
+            if haskey(model.name_to_var, var_name)
+                error("Variable $var and $(model.name_to_var[var_name]) have " *
```

Comment:
We should do `model.name_to_var = nothing` before erroring otherwise, we leave the model in an incorrect state